### DISCUSSION TITLE
How to compile the mono runtime to Wasm.

### DISCUSSION THREAD

**Discussion by shenqi-xiaoafei** [dotnet/runtime#95107] [Q&A] [answered]
I am using Windows 10, and the compile command is  “build.cmd mono.runtime -os browser -c Debug”

 Compilation exception:
[vcvarsall.bat] Environment initialized for: 'x86_x64'
  [1/2] Linking CXX executable mono\mini\mono-aot-cross.exe
  FAILED: mono/mini/mono-aot-cross.exe
  cmd.exe /C "cd . && "E:\Program Files\CMake\bin\cmake.exe" -E vs_link_exe --intdir=mono\mini\CMakeFiles\mono-sgen.dir --rc=C:\PROGRA~2\WI3CF2~1\10\bin\100220~1.0\x86\rc.exe --mt=C:\PROGRA~2\WI3CF2~1\10\bin\100220~1.0\x86\mt.exe --mani
  fests  -- E:\PROGRA~2\MICROS~1\2022\COMMUN~1\VC\Tools\MSVC\1431~1.311\bin\Hostx86\x64\link.exe /nologo @CMakeFiles\mono-sgen.rsp  /out:mono\mini\mono-aot-cross.exe /implib:mono\mini\mono-aot-cross.lib /pdb:mono\mini\mono-aot-cross.pdb
   /version:0.0 /machine:x64 /INCREMENTAL:NO /subsystem:console  /STACK:0x800000 /LTCG /DEBUG /OPT:REF /OPT:ICF /OPT:NOICF  && cd ."
  LINK: command "E:\PROGRA~2\MICROS~1\2022\COMMUN~1\VC\Tools\MSVC\1431~1.311\bin\Hostx86\x64\link.exe /nologo @CMakeFiles\mono-sgen.rsp /out:mono\mini\mono-aot-cross.exe /implib:mono\mini\mono-aot-cross.lib /pdb:mono\mini\mono-aot-cross
  .pdb /version:0.0 /machine:x64 /INCREMENTAL:NO /subsystem:console /STACK:0x800000 /LTCG /DEBUG /OPT:REF /OPT:ICF /OPT:NOICF /MANIFEST /MANIFESTFILE:mono\mini\mono-aot-cross.exe.manifest" failed (exit code 1120) with the following outp
  ut:
  LLVMCore.lib(SafepointIRVerifier.cpp.obj) : error LNK2001: 无法解析的外部符号 __std_init_once_link_alternate_names_and_abort
  LLVMAnalysis.lib(LoopAccessAnalysis.cpp.obj) : error LNK2001: 无法解析的外部符号 __std_init_once_link_alternate_names_and_abort
  LLVMCore.lib(Verifier.cpp.obj) : error LNK2001: 无法解析的外部符号 __std_init_once_link_alternate_names_and_abort
  LLVMCore.lib(Dominators.cpp.obj) : error LNK2001: 无法解析的外部符号 __std_init_once_link_alternate_names_and_abort
  LLVMCore.lib(IRPrintingPasses.cpp.obj) : error LNK2001: 无法解析的外部符号 __std_init_once_link_alternate_names_and_abort
  LLVMAnalysis.lib(TypeBasedAliasAnalysis.cpp.obj) : error LNK2001: 无法解析的外部符号 __std_init_once_link_alternate_names_and_abort
  LLVMAnalysis.lib(ProfileSummaryInfo.cpp.obj) : error LNK2001: 无法解析的外部符号 __std_init_once_link_alternate_names_and_abort
  LLVMAnalysis.lib(TargetLibraryInfo.cpp.obj) : error LNK2001: 无法解析的外部符号 __std_init_once_link_alternate_names_and_abort
  LLVMAnalysis.lib(AssumptionCache.cpp.obj) : error LNK2001: 无法解析的外部符号 __std_init_once_link_alternate_names_and_abort
  LLVMAnalysis.lib(ScalarEvolution.cpp.obj) : error LNK2001: 无法解析的外部符号 __std_init_once_link_alternate_names_and_abort
  LLVMAnalysis.lib(ScopedNoAliasAA.cpp.obj) : error LNK2001: 无法解析的外部符号 __std_init_once_link_alternate_names_and_abort
  LLVMAnalysis.lib(StackSafetyAnalysis.cpp.obj) : error LNK2001: 无法解析的外部符号 __std_init_once_link_alternate_names_and_abort
  LLVMAnalysis.lib(TargetTransformInfo.cpp.obj) : error LNK2001: 无法解析的外部符号 __std_init_once_link_alternate_names_and_abort
  LLVMAnalysis.lib(PostDominators.cpp.obj) : error LNK2001: 无法解析的外部符号 __std_init_once_link_alternate_names_and_abort
  LLVMAnalysis.lib(RegionInfo.cpp.obj) : error LNK2001: 无法解析的外部符号 __std_init_once_link_alternate_names_and_abort

**Comment by hez2010** [comment]
It seems that you don't have LLVM installed on your machine.

**Reply by shenqi-xiaoafei** [reply]
I have installed LLVM and set up the environment.

**Comment by maraf** [comment]
Here is some docs https://github.com/dotnet/runtime/blob/main/docs/workflow/building/mono/README.md with requirements for building on Windows. Also you should be building with `build.cmd -os browser -subset mono+libs -c Debug` to get all runtime dependencies built

**Reply by shenqi-xiaoafei** [reply]
Using this command, the same error occurs.

**Reply by maraf** [reply, marked as answer]
Do you have all the requirements from the docs? Can you try deleting `.\artifacts` and build from scratch?

**Reply by shenqi-xiaoafei** [reply]
I have successfully compiled WASM! I upgraded Viusal Studio 17.1 to 17.8, deleted the artifacts folder and rebuilt the project.
